### ISSUE TITLE
Add way to assign Entity IDs different from Entity Name

### ISSUE THREAD

**Issue by nanderson97651** [esphome/feature-requests#2119]
**Describe the problem you have/What new integration you would like**

Building off of #2114.  Add a way to separate the device and entity IDs from the friendly name.  This would allow an update to the friendly name of a device for re-use, but keep the entity ID constant.

**Please describe your use case for this integration and alternatives you've tried:**

Example:  I have a sonoff s31 outlet that is used for christmas lights in the winter and a fan in the summer.  Current implementation of friendly name will change all entity IDs forcing a refactor of dashboards and automations.  Allowing an option to set the entity name while leaving the entity ID intact (as can be done in the HA UI).  i.e. I move outlet from christmas lights to fan and have an optional flag to force ***entity_id*** to be derived from ***name*** rather than ***friendly name***, leaving the entity in HA as ***switch.sonoff_s31_relay*** with a name of ***Bedroom Fan Relay.***

![image](https://user-images.githubusercontent.com/73798513/219089524-08602e68-1069-4399-8e09-1a0ad1e3e5be.png)

This also becomes useful with HA 2023.2's ability to group sensor states.  I have a sensor group of energy consumed by all smart plugs to display in my energy dashboard but anytime I move a device its ID is updated forcing me to update the group and losing past data.

**Additional context**

This is admittedly a pretty narrow use case but I think in general the ability to disconnect entity ID from entity name without hand editing dozens of items could be useful.

**Comment by ezpl** [comment]
Your use case could be narrow, but the general problem is considerable. Not being able to change the display name without breaking things is not good design!

**Comment by EmhyrVarEmreis** [comment]
I was looking for this exact feature many times, and here it is, unimplemented :/

**Comment by ezpl** [comment]
> I was looking for this exact feature many times, and here it is, unimplemented :/

Yes, I was really surprised when first discovering this fragile design. Can't imagine it is difficult to add as an option. Tempted to look into it myself..

**Comment by abichinger** [comment]
**Proposal**
Add `object_id` as an optional variable to the [entity schema](https://github.com/esphome/esphome/blob/3eaf59cc5a411668ecb310cfb000c8ba85716c0b/esphome/config_validation.py#L1797). 

```yaml
# Example sensor configuration
sensor:
  name: Temperature
  object_id: temp_1
```

This would also require some changes [here](https://github.com/esphome/esphome/blob/3eaf59cc5a411668ecb310cfb000c8ba85716c0b/esphome/cpp_helpers.py#L102).

## Case 1: [Friendly Naming](https://www.home-assistant.io/integrations/esphome/#friendly-naming)

If `friendly_name` is set the resulting `entity_id` is [`f"{domain}.{device_info.name}_{entity_info.object_id}"`](https://github.com/home-assistant/core/blob/9689cb448db3ab5ce635d899fec3feacb4618875/homeassistant/components/esphome/entity.py#L204)

Example:

```yaml
esphome:
   name: "livingroom"
   friendly_name: "Living room"
```

The `entity_id` would be `sensor.livingroom_temp_1`

## Case 2: [Legacy Naming](https://www.home-assistant.io/integrations/esphome/#legacy-naming)

Example:

```yaml
esphome:
   name: "livingroom"
```

Without any changes to the esphome component of HA the entity ID would be `sensor.temperature`, because HA uses the [name of the sensor as object_id](https://github.com/home-assistant/core/blob/9689cb448db3ab5ce635d899fec3feacb4618875/homeassistant/helpers/entity.py#L698). To get the entity ID `sensor.temp_1` the [`EsphomeEntity.entity_id`]() could be set to `f"{domain}.{entity_info.object_id}"`

**Comment by jougs** [comment]
A solution to this problem would also be extremely useful in situations where the friendly name of the entity contains special characters.

**Here's an example**: I have a light in my *Gewölbekeller* (vaulted cellar), for which I would like the `name` to be `Gewölbekeller` (in German, with the umlaut), while I'd strongly like my `entity_id` to be `light.vaulted_cellar`.

Right now, the `entity_id` becomes `light.gewolbekeller`. This makes automations a mess due to a mix between the English entity names I use everywhere else and the (weirdly, it should be *gew**oe**lbekeller*) auto-transliterated ones coming from ESPHome.

**A fully backwards-compatible solution** would be add an optional field `friendly_name` for *all entities*.
* if unset, `name` in ESPHome would be used for both `entity_id` and `name` in Home Assistant, as it is now
* if set, `name` would become the `entity_id` in HA and `friendly_name` would become `name`.

**An alternative** could be to expose the `id` field of the entity as `entity_id` to Home Assistant if it is set and use the ESPHome entity `name` as `name` in HA. This behavior could be made optional by setting `internal` to `False` explicitly for the entity.

**Comment by clarencelocke** [comment]
+1 for getting this kind of change added.

I believe this situation would be correctly addressed by https://github.com/esphome/esphome/pull/4508

The whole crux of the matter here is that an entity's `object_id` in ESPHome is what is used as Home Assistant's entityID — and the **only** way it can be defined right now (see [here](https://github.com/esphome/esphome/blob/7edf45889806811c31295755300779b102110907/esphome/cpp_helpers.py#L102)) is as a **side effect** of the entity's `name` or the core `friendly_name`.

While having this automatically-produced property may be convenient for an ease of entry into using ESPHome in Home Assistant, it can so easily cause breakage of scripts from within Home Assistant if you decide you want a different / better name to be presented in the UI (i.e., you are forced to change the `name` parameter).  A programming identifier (the entity ID) should always be decoupled from it's human-readable representation (the name).

Having an optional `object_id` parameter in the yaml configuration which would take precedences over the (`friendly_`)`name`-based properties allows for explicit (and more importantly, _stable_) control over entity names while still remaining backward compatible with legacy behavior.

The only way I see to get around this without a PR is to add a lambda function to a `esphome:|on_boot:` entry which then calls `id(___).set_object_id("...");` for each entity involved.  Painful and ugly.

**Comment by tjakubo** [comment]
It's quite baffling that this hasn't been exposed after years and dozens of requests. Yes, usually the autogenerated ID is good. But there's plenty of cases when not being able to control it *at all* is a huge pain. A simple optional property that doesn't try to be smart and simply sets the ID would be perfect for that.

**Comment by cravl-dev** [comment]
The issue with Unicode collisions mentioned by @Ziron is now fixed (at least in the web server) as of [2026.1.0](https://www.esphome.io/changelog/2026.1.0/#web-server-improvements).

However, the issue of not being able to override auto-generated IDs is still a **big** oversight IMO. I haven't worked on ESPHome stuff in a couple months because I designed a whole new config structure around the *assumption* that I could set something as basic as entity IDs manually (using Jinja expressions to create a consistent schema that makes sense to me). 

Unfortunately, instead of what the Jinja outputs:
  `sensor.bedroom_atmo_node_uptime`

...it still tries to be smart and I end up with entities with IDs like:
  `sensor.node_bedroom_dresser_atmo_node_uptime`
*how even... the heck?*

Then, if I rearrange the furniture and that node ends up not on the dresser anymore, I'll want to rename it, which messes up all the IDs and I have to do a careful regex search-and-replace through my entire config (and sensor history!) over SSH.

**Comment by jougs** [comment]
See also https://github.com/home-assistant/core/issues/97558
